src/main/scala/chisel3/util/SRAM.scala

```diff
@@ -0,0 +1,255 @@
+    )
+  }
+  override def typeName: String = s"SRAMInterface_${SRAM.portedness(numReadPorts, numWritePorts, numReadwritePorts)}"
+
+  val readPorts:  Vec[MemoryReadPort[T]] = Vec(numReadPorts, new MemoryReadPort(tpe, addrWidth))
+  val writePorts: Vec[MemoryWritePort[T]] = Vec(numWritePorts, new MemoryWritePort(tpe, addrWidth, masked))
+  val readwritePorts: Vec[MemoryReadWritePort[T]] =
+    Vec(numReadwritePorts, new MemoryReadWritePort(tpe, addrWidth, masked))
+}
+
+object SRAM {
+
+  /** Generates a [[SyncReadMem]] within the current module, connected to an explicit number
+    * of read, write, and read/write ports
+    *
+    * @param size The desired size of the inner `SyncReadMem`
+    * @tparam T The data type of the memory element
+    * @param numReadPorts The number of desired read ports, >= 0
+    * @param numWritePorts The number of desired write ports, >= 0
+    * @param numReadwritePorts The number of desired read/write ports, >= 0
+    *
+    * @return A new `SRAMInterface` wire containing the control signals for each instantiated port
+    * @note This does *not* return the `SyncReadMem` itself, you must interact with it using the returned bundle
+    */
+  def apply[T <: Data](
+    size:              BigInt,
+    tpe:               T,
+    numReadPorts:      Int,
+    numWritePorts:     Int,
+    numReadwritePorts: Int
+  )(
+    implicit sourceInfo: SourceInfo
+  ): SRAMInterface[T] =
+    memInterface_impl(size, tpe)(numReadPorts, numWritePorts, numReadwritePorts, Builder.forcedClock)
+
+  /** Generates a [[SyncReadMem]] within the current module, connected to an explicit number
+    * of read, write, and read/write ports, with masking capability on all write and read/write ports.
+    *
+    * @param size The desired size of the inner `SyncReadMem`
+    * @tparam T The data type of the memory element
+    * @param numReadPorts The number of desired read ports, >= 0
+    * @param numWritePorts The number of desired write ports, >= 0
+    * @param numReadwritePorts The number of desired read/write ports, >= 0
+    *
+    * @return A new `SRAMInterface` wire containing the control signals for each instantiated port
+    * @note This does *not* return the `SyncReadMem` itself, you must interact with it using the returned bundle
+    */
+  def masked[T <: Data](
+    size:              BigInt,
+    tpe:               T,
+    numReadPorts:      Int,
+    numWritePorts:     Int,
+    numReadwritePorts: Int
+  )(
+    implicit evidence: T <:< Vec[_],
+    sourceInfo:        SourceInfo
+  ): SRAMInterface[T] =
+    memInterface_impl(size, tpe)(numReadPorts, numWritePorts, numReadwritePorts, Builder.forcedClock)
+
+  /** @group SourceInfoTransformMacro */
+  private def memInterface_impl[T <: Data](
+    size:              BigInt,
+    tpe:               T
+  )(numReadPorts:      Int,
+    numWritePorts:     Int,
+    numReadwritePorts: Int,
+    clock:             Clock
```

Comment:
I have been thinking: what's this `clock` used for.
I think `clock` should always be bind to port rather than a SRAM instance?

Reply:
I agree we don't need this particular clock, but I have a different opinion on what the clock API should look like: https://github.com/chipsalliance/chisel/pull/3228/files#r1220340613